File: desktop/shared/planning.ts
import type { AnswerInputRequest, InputAnswer, InputQuestion, PlanDocument, PlanEditRequest, PlanningReadResponse, RunPlanningOptions, UserInputRequest, UserInputResponse } from "./planning.generated.js";

function record(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}
function keys(value: Record<string, unknown>, allowed: string[]): boolean {
  return Object.keys(value).every((key) => allowed.includes(key));
}
function text(value: unknown, max: number, min = 0): value is string {
  return typeof value === "string" && value.length >= min && value.length <= max;
}
function positive(value: unknown): value is number {
  return Number.isSafeInteger(value) && Number(value) > 0;
}
function answer(value: unknown): value is InputAnswer {
  return record(value) && keys(value, ["questionId", "optionIds", "text"])
    && text(value.questionId, 80, 1) && (value.text === undefined || text(value.text, 8000))
    && (value.optionIds === undefined || (Array.isArray(value.optionIds) && value.optionIds.length <= 6 && value.optionIds.every((id) => text(id, 80, 1))));
}
function response(value: unknown): value is UserInputResponse {
  return record(value) && keys(value, ["status", "answers"])
    && ["answered", "skipped"].includes(String(value.status))
    && (value.answers === undefined || (Array.isArray(value.answers) && value.answers.length <= 3 && value.answers.every(answer)));
}
function question(value: unknown): value is InputQuestion {
  return record(value) && keys(value, ["id", "question", "type", "options", "recommendedOptionId"])
    && text(value.id, 80, 1) && text(value.question, 1000, 1)
    && (value.type === undefined || ["single_select", "multi_select", "text"].includes(String(value.type)))
    && (value.recommendedOptionId == null || text(value.recommendedOptionId, 80, 1))
    && (value.options === undefined || (Array.isArray(value.options) && value.options.length <= 6 && value.options.every((option) =>
      record(option) && keys(option, ["id", "label", "description"]) && text(option.id, 80, 1)
      && text(option.label, 200, 1) && (option.description === undefined || text(option.description, 500)))));
}
export function isUserInputRequest(value: unknown): value is UserInputRequest {
  return record(value) && keys(value, ["id", "sessionId", "runId", "itemId", "questions", "status", "response", "createdAt"])
    && [value.id, value.sessionId, value.runId, value.itemId].every((id) => text(id, 128, 1))
    && Array.isArray(value.questions) && value.questions.length >= 1 && value.questions.length <= 3 && value.questions.every(question)
    && ["pending", "answered", "skipped", "canceled"].includes(String(value.status))
    && (value.response == null || response(value.response)) && positive(value.createdAt);
}
export function isPlanDocument(value: unknown): value is PlanDocument {
  return record(value) && keys(value, ["id", "sessionId", "runId", "revision", "title", "markdown", "sha256", "status", "path", "updatedAt", "executionRunId"])
    && [value.id, value.sessionId, value.runId].every((id) => text(id, 128, 1))
    && positive(value.revision) && text(value.title, 200, 1) && text(value.markdown, 65536, 1)
    && typeof value.sha256 === "string" && /^[a-f0-9]{64}$/.test(value.sha256)
    && ["draft", "review", "accepted"].includes(String(value.status)) && text(value.path, 4096, 1)
    && positive(value.updatedAt) && (value.executionRunId == null || text(value.executionRunId, 128, 1));
}
export function isPlanningReadResponse(value: unknown): value is PlanningReadResponse {
  return record(value) && keys(value, ["plans", "questions"])
    && Array.isArray(value.plans) && value.plans.length <= 50 && value.plans.every(isPlanDocument)
    && Array.isArray(value.questions) && value.questions.length <= 50 && value.questions.every(isUserInputRequest);
}
export function isPlanResponse(value: unknown): value is { plan: PlanDocument } {
  return record(value) && keys(value, ["plan"]) && isPlanDocument(value.plan);
}
export function isAnswerInputRequest(value: unknown): value is AnswerInputRequest {
  return record(value) && keys(value, ["requestId", "response"]) && text(value.requestId, 128, 1) && response(value.response);
}
export function isPlanEditRequest(value: unknown): value is PlanEditRequest {
  return record(value) && keys(value, ["planId", "expectedRevision", "markdown"])
    && text(value.planId, 128, 1) && positive(value.expectedRevision) && text(value.markdown, 65536, 1);
}
export function isRunPlanningOptions(value: unknown): value is RunPlanningOptions {
  return record(value) && keys(value, ["workMode", "planId", "planRevision"])
    && (value.workMode === undefined || ["execute", "plan"].includes(String(value.workMode)))
    && (value.planId == null || text(value.planId, 128, 1)) && (value.planRevision == null || positive(value.planRevision))
    && ((value.planId == null) === (value.planRevision == null));
}

File: runtime/eidos_runtime/domain/planning.py
from __future__ import annotations

from typing import Literal
from pydantic import Field, model_validator
from eidos_runtime.models import EidosFrozenStrictModel, JsonSafeInt

WorkMode = Literal['execute', 'plan']


class InputOption(EidosFrozenStrictModel):
    id: str = Field(min_length=1, max_length=80)
    label: str = Field(min_length=1, max_length=200)
    description: str = Field(default='', max_length=500)


class InputQuestion(EidosFrozenStrictModel):
    id: str = Field(min_length=1, max_length=80)
    question: str = Field(min_length=1, max_length=1000)
    type: Literal['single_select', 'multi_select', 'text'] = 'single_select'
    options: list[InputOption] = Field(default_factory=list, max_length=6)
    recommended_option_id: str | None = None

    @model_validator(mode='after')
    def validate_options(self):
        ids = [option.id for option in self.options]
        if len(ids) != len(set(ids)):
            raise ValueError('duplicate_option_id')
        if self.type == 'text' and self.options:
            raise ValueError('text_question_has_options')
        if self.type != 'text' and len(ids) < 2:
            raise ValueError('choice_question_requires_options')
        if self.recommended_option_id is not None and self.recommended_option_id not in ids:
            raise ValueError('unknown_recommended_option')
        return self


class RequestUserInput(EidosFrozenStrictModel):
    questions: list[InputQuestion] = Field(min_length=1, max_length=3)

    @model_validator(mode='after')
    def unique_questions(self):
        if len({q.id for q in self.questions}) != len(self.questions):
            raise ValueError('duplicate_question_id')
        return self


class InputAnswer(EidosFrozenStrictModel):
    question_id: str = Field(min_length=1, max_length=80)
    option_ids: list[str] = Field(default_factory=list, max_length=6)
    text: str = Field(default='', max_length=8000)


class UserInputResponse(EidosFrozenStrictModel):
    status: Literal['answered', 'skipped']
    answers: list[InputAnswer] = Field(default_factory=list, max_length=3)


class UserInputRequest(EidosFrozenStrictModel):
    id: str
    session_id: str
    run_id: str
    item_id: str
    questions: list[InputQuestion]
    status: Literal['pending', 'answered', 'skipped', 'canceled']
    response: UserInputResponse | None = None
    created_at: JsonSafeInt


class PlanDocument(EidosFrozenStrictModel):
    id: str
    session_id: str
    run_id: str
    revision: int = Field(ge=1)
    title: str = Field(min_length=1, max_length=200)
    markdown: str = Field(min_length=1, max_length=65536)
    sha256: str
    status: Literal['draft', 'review', 'accepted']
    path: str
    updated_at: JsonSafeInt
    execution_run_id: str | None = None


class WritePlan(EidosFrozenStrictModel):
    plan_id: str | None = None
    expected_revision: int | None = Field(default=None, ge=1)
    title: str = Field(min_length=1, max_length=200)
    markdown: str = Field(min_length=1, max_length=65536)
    ready_for_review: bool = False


class PlanningSuspended(Exception):
    """A durable user-input request owns the continuation; no worker waits."""
